crates/datasources/src/clickhouse/convert.rs

```diff
@@ -0,0 +1,334 @@
+        DataType::UInt32 => make_primitive_array!(UInt32, UInt32Array, nullable),
+        DataType::UInt64 => make_primitive_array!(UInt64, UInt64Array, nullable),
+        DataType::Int8 => make_primitive_array!(Int8, Int8Array, nullable),
+        DataType::Int16 => make_primitive_array!(Int16, Int16Array, nullable),
+        DataType::Int32 => make_primitive_array!(Int32, Int32Array, nullable),
+        DataType::Int64 => make_primitive_array!(Int64, Int64Array, nullable),
+        DataType::Float32 => make_primitive_array!(Float32, Float32Array, nullable),
+        DataType::Float64 => make_primitive_array!(Float64, Float64Array, nullable),
+        DataType::Utf8 => {
+            if nullable {
+                let mut vals = Vec::with_capacity(column.len());
+                for val in column {
+                    match val {
+                        Value::String(v) => vals.push(Some(String::from_utf8(v)?)),
+                        Value::Null if nullable => vals.push(None),
+                        other => {
+                            return Err(ClickhouseError::String(format!(
+                                "unexpected value type: {other}"
+                            )))
+                        }
+                    }
+                }
+                Arc::new(StringArray::from(vals))
+            } else {
+                let mut vals = Vec::with_capacity(column.len());
+                for val in column {
+                    match val {
+                        Value::String(v) => vals.push(String::from_utf8(v)?),
+                        other => {
+                            return Err(ClickhouseError::String(format!(
+                                "unexpected value type: {other}"
+                            )))
+                        }
+                    }
+                }
+                Arc::new(StringArray::from(vals))
+            }
+        }
+        DataType::Date32 => {
+            let epoch = NaiveDate::from_ymd_opt(1970, 1, 1).unwrap();
+            if nullable {
+                let mut vals: Vec<Option<NaiveDate>> = Vec::with_capacity(column.len());
+                for val in column {
+                    match val {
+                        Value::Date(v) => vals.push(Some(v.into())),
+                        Value::Null if nullable => vals.push(None),
+                        other => {
+                            return Err(ClickhouseError::String(format!(
+                                "unexpected value type: {other}"
+                            )))
+                        }
+                    }
+                }
+                Arc::new(Date32Array::from(
+                    vals.into_iter()
+                        .map(|date| {
+                            date.map(|date| date.signed_duration_since(epoch).num_days() as i32)
+                        })
+                        .collect::<Vec<_>>(),
```

Comment:
using the appropriate `ArrayBuilder`s would be faster & use less allocations than using an intermediate `Vec`?

Reply:
fixed